Batch matrix multiplication test fails for MPS device on float16 precision

### Expected Behavioral

Test suite to run perfect once I got my environment setup.



### System Info

I am using Python 3.11.7. This is my system information:

```
uname -a
Darwin Voyager 22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000 arm64
```

These are the Python environment:

```
% pip list
Package           Version     Editable project location
----------------- ----------- -----------------------------------
black             24.2.0
click             8.1.7
filelock          3.13.1
fsspec            2024.2.0
iniconfig         2.0.0
Jinja2            3.1.3
MarkupSafe        2.1.5
mpmath            1.3.0
mypy-extensions   1.0.0
networkx          3.2.1
ninja             1.11.1.1
packaging         23.2
pathspec          0.12.1
pip               23.2.1
platformdirs      4.2.0
pluggy            1.4.0
pytest            8.0.2
quanto            0.0.14.dev0 /Users/alejandro/Programming/quanto
ruff              0.3.0
setuptools        65.5.0
sympy             1.12
torch             2.2.1
typing_extensions 4.10.0
```

### Bug Logs

Here are the logs associated with `python -m pytest -sv test/tensor/ops/test_mm_dispatch.py`:

```
========================================================================= FAILURES =========================================================================
___________________________________________________ test_bmm[mps-b_per_tensor-a_shape0-b_shape0-1-fp16] ____________________________________________________

dtype = torch.float16, batch_size = 1, a_shape = (16, 32), b_shape = (32, 24), b_axis = None, device = device(type='mps')

    @pytest.mark.parametrize("dtype", [torch.float32, torch.float16], ids=["fp32", "fp16"])
    @pytest.mark.parametrize("batch_size", [1, 10])
    @pytest.mark.parametrize("a_shape, b_shape", [[(16, 32), (32, 24)], [(5, 10), (10, 6)]])
    @pytest.mark.parametrize("b_axis", [None, 2], ids=["b_per_tensor", "b_per_axis"])
    def test_bmm(dtype, batch_size, a_shape, b_shape, b_axis, device):
        if dtype == torch.float16 and device.type == "cpu":
            pytest.skip("Matrix multiplication is not supported for float16 on CPU.")
        qa = random_qtensor((batch_size,) + a_shape, dtype=dtype).to(device)
        qb = random_qtensor((batch_size,) + b_shape, axis=b_axis, dtype=dtype).to(device)
        qbmm = torch.bmm(qa, qb)
        # The outputs should be almost identical if we use the dequantized inputs
        bmm = torch.bmm(qa.dequantize(), qb.dequantize())
        # We need to increase atol and rtol for float16
        atol = {torch.float32: 1e-6, torch.float16: 2e-3}[dtype]
        rtol = {torch.float32: 1e-5, torch.float16: 1e-2}[dtype]
>       assert_close(bmm, qbmm, atol=atol, rtol=rtol)

test/tensor/ops/test_mm_dispatch.py:41: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = tensor([[[ 4.1719e+00,  5.7861e-02, -8.2715e-01, -7.7832e-01, -8.3496e-02,
          -6.5820e-01, -7.9004e-01,  6.6846....9961e+00,
           3.7988e+00, -2.2637e+00, -1.3203e+00,  2.8457e+00]]],
       device='mps:0', dtype=torch.float16)
b = tensor([[[ 4.1758e+00,  5.7617e-02, -8.2861e-01, -7.7930e-01, -8.2642e-02,
          -6.5723e-01, -7.8760e-01,  6.6846....9941e+00,
           3.7988e+00, -2.2637e+00, -1.3193e+00,  2.8457e+00]]],
       device='mps:0', dtype=torch.float16)
atol = tensor(0.0020), rtol = 0.01

    def assert_close(a: torch.Tensor, b: torch.Tensor, atol: float = None, rtol: float = None):
        # Please refer to that discussion for default rtol values based on the float type:
        # https://scicomp.stackexchange.com/questions/43111/float-equality-tolerance-for-single-and-half-precision
        assert a.dtype == b.dtype
        if atol is None:
            # We use torch finfo resolution
            atol = torch.finfo(a.dtype).resolution
        atol = torch.tensor(atol)
        # We cannot expect an absolute error lower than the quantization scales
        if isinstance(a, QTensor):
            atol = torch.maximum(a._scale, atol)
            a = a.dequantize()
        if isinstance(b, QTensor):
            atol = torch.maximum(b._scale, atol)
            b = b.dequantize()
        if rtol is None:
            # Please refer to that discussion for default rtol values based on the float type:
            # https://scicomp.stackexchange.com/questions/43111/float-equality-tolerance-for-single-and-half-precision
            rtol = {torch.float32: 1e-5, torch.float16: 1e-3, torch.bfloat16: 1e-1}[a.dtype]
        abs_error = torch.abs(a - b)
        closeness = atol + rtol * torch.abs(a)
        if not torch.all(abs_error <= closeness):
            max_error_index = torch.argmax(abs_error - closeness)
            max_error_a = a.flatten()[max_error_index]
            max_rel_error = abs_error.flatten()[max_error_index] / torch.abs(max_error_a) * 100
            max_error_b = b.flatten()[max_error_index]
>           raise ValueError(
                f"Error exceeds tolerance (max: {max_error_b:.8f} instead of {max_error_a:.8f} ({max_rel_error:.4f} %)."
            )
E           ValueError: Error exceeds tolerance (max: 0.01824951 instead of 0.02050781 (11.0078 %).

test/helpers.py:73: ValueError
___________________________________________________ test_bmm[mps-b_per_tensor-a_shape0-b_shape0-10-fp16] ___________________________________________________

dtype = torch.float16, batch_size = 10, a_shape = (16, 32), b_shape = (32, 24), b_axis = None, device = device(type='mps')

    @pytest.mark.parametrize("dtype", [torch.float32, torch.float16], ids=["fp32", "fp16"])
    @pytest.mark.parametrize("batch_size", [1, 10])
    @pytest.mark.parametrize("a_shape, b_shape", [[(16, 32), (32, 24)], [(5, 10), (10, 6)]])
    @pytest.mark.parametrize("b_axis", [None, 2], ids=["b_per_tensor", "b_per_axis"])
    def test_bmm(dtype, batch_size, a_shape, b_shape, b_axis, device):
        if dtype == torch.float16 and device.type == "cpu":
            pytest.skip("Matrix multiplication is not supported for float16 on CPU.")
        qa = random_qtensor((batch_size,) + a_shape, dtype=dtype).to(device)
        qb = random_qtensor((batch_size,) + b_shape, axis=b_axis, dtype=dtype).to(device)
        qbmm = torch.bmm(qa, qb)
        # The outputs should be almost identical if we use the dequantized inputs
        bmm = torch.bmm(qa.dequantize(), qb.dequantize())
        # We need to increase atol and rtol for float16
        atol = {torch.float32: 1e-6, torch.float16: 2e-3}[dtype]
        rtol = {torch.float32: 1e-5, torch.float16: 1e-2}[dtype]
>       assert_close(bmm, qbmm, atol=atol, rtol=rtol)

test/tensor/ops/test_mm_dispatch.py:41: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = tensor([[[ 6.0254e-01,  9.0430e-01,  1.8420e-01,  ...,  1.2881e+00,
          -2.1504e+00, -4.9829e-01],
         [ 1....2.3379e+00,  1.0078e+00,  ..., -2.5879e-01,
          -7.2266e-02, -1.0439e+00]]], device='mps:0', dtype=torch.float16)
b = tensor([[[ 6.0352e-01,  9.0479e-01,  1.8652e-01,  ...,  1.2900e+00,
          -2.1504e+00, -4.9756e-01],
         [ 1....2.3379e+00,  1.0078e+00,  ..., -2.5879e-01,
          -7.2083e-02, -1.0449e+00]]], device='mps:0', dtype=torch.float16)
atol = tensor(0.0020), rtol = 0.01

    def assert_close(a: torch.Tensor, b: torch.Tensor, atol: float = None, rtol: float = None):
        # Please refer to that discussion for default rtol values based on the float type:
        # https://scicomp.stackexchange.com/questions/43111/float-equality-tolerance-for-single-and-half-precision
        assert a.dtype == b.dtype
        if atol is None:
            # We use torch finfo resolution
            atol = torch.finfo(a.dtype).resolution
        atol = torch.tensor(atol)
        # We cannot expect an absolute error lower than the quantization scales
        if isinstance(a, QTensor):
            atol = torch.maximum(a._scale, atol)
            a = a.dequantize()
        if isinstance(b, QTensor):
            atol = torch.maximum(b._scale, atol)
            b = b.dequantize()
        if rtol is None:
            # Please refer to that discussion for default rtol values based on the float type:
            # https://scicomp.stackexchange.com/questions/43111/float-equality-tolerance-for-single-and-half-precision
            rtol = {torch.float32: 1e-5, torch.float16: 1e-3, torch.bfloat16: 1e-1}[a.dtype]
        abs_error = torch.abs(a - b)
        closeness = atol + rtol * torch.abs(a)
        if not torch.all(abs_error <= closeness):
            max_error_index = torch.argmax(abs_error - closeness)
            max_error_a = a.flatten()[max_error_index]
            max_rel_error = abs_error.flatten()[max_error_index] / torch.abs(max_error_a) * 100
            max_error_b = b.flatten()[max_error_index]
>           raise ValueError(
                f"Error exceeds tolerance (max: {max_error_b:.8f} instead of {max_error_a:.8f} ({max_rel_error:.4f} %)."
            )
E           ValueError: Error exceeds tolerance (max: 0.02473450 instead of 0.02221680 (11.3359 %).

test/helpers.py:73: ValueError
____________________________________________________ test_bmm[mps-b_per_axis-a_shape0-b_shape0-10-fp16] ____________________________________________________

dtype = torch.float16, batch_size = 10, a_shape = (16, 32), b_shape = (32, 24), b_axis = 2, device = device(type='mps')

    @pytest.mark.parametrize("dtype", [torch.float32, torch.float16], ids=["fp32", "fp16"])
    @pytest.mark.parametrize("batch_size", [1, 10])
    @pytest.mark.parametrize("a_shape, b_shape", [[(16, 32), (32, 24)], [(5, 10), (10, 6)]])
    @pytest.mark.parametrize("b_axis", [None, 2], ids=["b_per_tensor", "b_per_axis"])
    def test_bmm(dtype, batch_size, a_shape, b_shape, b_axis, device):
        if dtype == torch.float16 and device.type == "cpu":
            pytest.skip("Matrix multiplication is not supported for float16 on CPU.")
        qa = random_qtensor((batch_size,) + a_shape, dtype=dtype).to(device)
        qb = random_qtensor((batch_size,) + b_shape, axis=b_axis, dtype=dtype).to(device)
        qbmm = torch.bmm(qa, qb)
        # The outputs should be almost identical if we use the dequantized inputs
        bmm = torch.bmm(qa.dequantize(), qb.dequantize())
        # We need to increase atol and rtol for float16
        atol = {torch.float32: 1e-6, torch.float16: 2e-3}[dtype]
        rtol = {torch.float32: 1e-5, torch.float16: 1e-2}[dtype]
>       assert_close(bmm, qbmm, atol=atol, rtol=rtol)

test/tensor/ops/test_mm_dispatch.py:41: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

a = tensor([[[ 0.9648, -0.5015, -0.9287,  ..., -2.0859, -1.0752,  0.9463],
         [-2.1973, -2.2754, -3.1543,  ...,  1.7...],
         [-1.6230,  0.2109,  0.8936,  ..., -1.1660, -1.5488,  3.3770]]],
       device='mps:0', dtype=torch.float16)
b = tensor([[[ 0.9648, -0.5005, -0.9297,  ..., -2.0840, -1.0762,  0.9473],
         [-2.1992, -2.2773, -3.1543,  ...,  1.7...],
         [-1.6240,  0.2109,  0.8940,  ..., -1.1660, -1.5498,  3.3750]]],
       device='mps:0', dtype=torch.float16)
atol = tensor(0.0020), rtol = 0.01

    def assert_close(a: torch.Tensor, b: torch.Tensor, atol: float = None, rtol: float = None):
        # Please refer to that discussion for default rtol values based on the float type:
        # https://scicomp.stackexchange.com/questions/43111/float-equality-tolerance-for-single-and-half-precision
        assert a.dtype == b.dtype
        if atol is None:
            # We use torch finfo resolution
            atol = torch.finfo(a.dtype).resolution
        atol = torch.tensor(atol)
        # We cannot expect an absolute error lower than the quantization scales
        if isinstance(a, QTensor):
            atol = torch.maximum(a._scale, atol)
            a = a.dequantize()
        if isinstance(b, QTensor):
            atol = torch.maximum(b._scale, atol)
            b = b.dequantize()
        if rtol is None:
            # Please refer to that discussion for default rtol values based on the float type:
            # https://scicomp.stackexchange.com/questions/43111/float-equality-tolerance-for-single-and-half-precision
            rtol = {torch.float32: 1e-5, torch.float16: 1e-3, torch.bfloat16: 1e-1}[a.dtype]
        abs_error = torch.abs(a - b)
        closeness = atol + rtol * torch.abs(a)
        if not torch.all(abs_error <= closeness):
            max_error_index = torch.argmax(abs_error - closeness)
            max_error_a = a.flatten()[max_error_index]
            max_rel_error = abs_error.flatten()[max_error_index] / torch.abs(max_error_a) * 100
            max_error_b = b.flatten()[max_error_index]
>           raise ValueError(
                f"Error exceeds tolerance (max: {max_error_b:.8f} instead of {max_error_a:.8f} ({max_rel_error:.4f} %)."
            )
E           ValueError: Error exceeds tolerance (max: 0.04104614 instead of 0.04418945 (7.1094 %).

test/helpers.py:73: ValueError
===================================================================== warnings summary =====================================================================
quanto_dev/lib/python3.11/site-packages/torch/nn/modules/transformer.py:20
  /Users/alejandro/Programming/quanto/quanto_dev/lib/python3.11/site-packages/torch/nn/modules/transformer.py:20: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/torch/csrc/utils/tensor_numpy.cpp:84.)
    device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================= short test summary info ==================================================================
FAILED test/tensor/ops/test_mm_dispatch.py::test_bmm[mps-b_per_tensor-a_shape0-b_shape0-1-fp16] - ValueError: Error exceeds tolerance (max: 0.01824951 instead of 0.02050781 (11.0078 %).
FAILED test/tensor/ops/test_mm_dispatch.py::test_bmm[mps-b_per_tensor-a_shape0-b_shape0-10-fp16] - ValueError: Error exceeds tolerance (max: 0.02473450 instead of 0.02221680 (11.3359 %).
FAILED test/tensor/ops/test_mm_dispatch.py::test_bmm[mps-b_per_axis-a_shape0-b_shape0-10-fp16] - ValueError: Error exceeds tolerance (max: 0.04104614 instead of 0.04418945 (7.1094 %).
```

--- a/test/tensor/ops/test_linear_dispatch.py
+++ b/test/tensor/ops/test_linear_dispatch.py
@@ -1,6 +1,6 @@
 import pytest
 import torch
-from helpers import assert_close, random_qtensor, random_tensor
+from helpers import assert_similar, random_qtensor, random_tensor
 
 
 @pytest.mark.parametrize("batch_size", [1, 10])
@@ -16,6 +16,4 @@ def test_linear(batch_size, tokens, embeddings, use_bias, dtype, weight_axis, de
     bias = random_tensor((embeddings,), dtype=dtype).to(device) if use_bias else None
     out = torch.nn.functional.linear(qinputs.dequantize(), qweight.dequantize(), bias)
     qout = torch.nn.functional.linear(qinputs, qweight, bias)
-    # We need to increase rtol for float16
-    rtol = {torch.float32: 1e-5, torch.float16: 1e-2}[dtype]
-    assert_close(out, qout, rtol)
+    assert_similar(out, qout)

--- a/test/tensor/ops/test_mm_dispatch.py
+++ b/test/tensor/ops/test_mm_dispatch.py
@@ -1,6 +1,6 @@
 import pytest
 import torch
-from helpers import assert_close, random_qtensor, random_tensor
+from helpers import assert_similar, random_qtensor, random_tensor
 
 from quanto import QTensor
 
@@ -17,10 +17,7 @@ def test_matmul(dtype, in_features, hidden, out_features, device):
     qmatmul = torch.matmul(qa, qb)
     # The outputs should be almost identical if we use the dequantized inputs
     matmul = torch.matmul(qa.dequantize(), qb.dequantize())
-    # We need to increase atol and rtol for float16
-    atol = {torch.float32: 1e-6, torch.float16: 2e-3}[dtype]
-    rtol = {torch.float32: 1e-5, torch.float16: 1e-2}[dtype]
-    assert_close(matmul, qmatmul, atol=atol, rtol=rtol)
+    assert_similar(matmul, qmatmul)
 
 
 @pytest.mark.parametrize("dtype", [torch.float32, torch.float16], ids=["fp32", "fp16"])
@@ -35,10 +32,7 @@ def test_bmm(dtype, batch_size, a_shape, b_shape, b_axis, device):
     qbmm = torch.bmm(qa, qb)
     # The outputs should be almost identical if we use the dequantized inputs
     bmm = torch.bmm(qa.dequantize(), qb.dequantize())
-    # We need to increase atol and rtol for float16
-    atol = {torch.float32: 1e-6, torch.float16: 2e-3}[dtype]
-    rtol = {torch.float32: 1e-5, torch.float16: 1e-2}[dtype]
-    assert_close(bmm, qbmm, atol=atol, rtol=rtol)
+    assert_similar(bmm, qbmm)
 
 
 @pytest.mark.parametrize(

--- a/test/tensor/ops/test_qtensor_dispatch.py
+++ b/test/tensor/ops/test_qtensor_dispatch.py
@@ -1,6 +1,6 @@
 import pytest
 import torch
-from helpers import assert_close, random_qtensor, random_tensor
+from helpers import assert_similar, random_qtensor, random_tensor
 
 from quanto import QTensor
 
@@ -23,11 +23,11 @@ def test_mul_scalar(input_shape, scalar, device):
     qprod = qa * scalar
     assert isinstance(qprod, QTensor)
     prod = qa.dequantize() * scalar
-    assert_close(prod, qprod)
+    assert_similar(prod, qprod)
     qprod = scalar * qa
     assert isinstance(qprod, QTensor)
     prod = scalar * qa.dequantize()
-    assert_close(prod, qprod)
+    assert_similar(prod, qprod)
 
 
 @pytest.mark.parametrize("batch_size", [1, 10])
@@ -64,7 +64,7 @@ def test_cat(input_shape, device):
     qother = QTensor.quantize(other, qtype=qinputs.qtype, axis=None, group_size=None, scale=qinputs._scale)
     qcat = torch.cat([qinputs, qother])
     assert isinstance(qcat, QTensor)
-    assert_close(torch.cat([qinputs.dequantize(), qother.dequantize()]), qcat)
+    assert_similar(torch.cat([qinputs.dequantize(), qother.dequantize()]), qcat)
     # Now, verify that with different scales, the output is dequantized
     qother = QTensor.quantize(other, qinputs.qtype)
     qcat = torch.cat([qinputs, qother])

